kernel/libc/koslib/gmon.c

```diff
@@ -0,0 +1,386 @@
+                index = (pc - cxt->lowpc) / HISTFRACTION;
+
+                /* Increment the histogram counter */
+                cxt->histogram[index]++;
+            }
+
+            /* Sleep for 10ms before the next sample */
+            thd_sleep(HISTOGRAM_INTERVAL_MS);
+        }
+
+    }
+
+    return NULL;
+}
+
+void moncontrol(bool mode) {
+    gmon_context_t *cxt = &g_context;
+
+    /* Don't change the state if we ran into an error */
+    if(cxt->state == GMON_PROF_ERROR)
+        return;
+
+    /* Treat start request as stop if error or gmon not initialized */
+    if(mode && cxt->initialized) {
+        /* Start */
+        cxt->state = GMON_PROF_ON;
+    } else {
+        /* Stop */
+        cxt->state = GMON_PROF_OFF;
+    }
+}
+
+/* Called each time we enter a function */
+void _mcount(uintptr_t frompc, uintptr_t selfpc) {
+    uint32_t index;
+    uint16_t node_index;
+    uint16_t *index_ptr;
+    gmon_node_t *node;
+    gmon_context_t *cxt = &g_context;
+
+    if(cxt->state != GMON_PROF_ON)
+        return;
+
+    /* If function is within the .text section */
+    if(frompc >= cxt->lowpc && frompc <= cxt->highpc) {
+        cxt->pc = selfpc; /* Save the last place we are */
+
+        /* Calculate index into 'froms' array */
+        index = (frompc - cxt->lowpc) / HASHFRACTION;
+        index_ptr = &cxt->froms[index];
+
+        /* Grab index into node array */
+	    node_index = *index_ptr;
+
+        /* Node doesnt exist? */
+        if(node_index == 0) {
+            goto create_node;
+        }
+
+        /* Try and find the node */
+        node = &cxt->nodes[node_index];
+
+        while(true) {
+            /* Found the node */
+            if(node->selfpc == selfpc) {
+                node->count++;
+                return;
+            }
+            /* Check if selfpc is less */
+            else if(selfpc < node->selfpc) {
+                if(node->left == 0) {
+                    index_ptr = &node->left;
+                    goto create_node;
+                }
+
+                node = &cxt->nodes[node->left];
+            }
+            else {
+                if(node->right == 0) {
+                    index_ptr = &node->right;
+                    goto create_node;
+                }
+
+                node = &cxt->nodes[node->right];
+            }
+        }
+
+create_node:
+        node_index = ++cxt->node_count;
+        if(node_index >= cxt->nnodes)
+            /* halt further profiling */
+            goto overflow;
+
+        *index_ptr = node_index;
+        node = &cxt->nodes[node_index];
+        node->selfpc = selfpc;
+        node->count = 1;
+        node->left = 0;
+        node->right = 0;
+        return;
+
+overflow:
+        cxt->state = GMON_PROF_ERROR;
+        //ERR("mcount: call graph buffer size limit exceeded, gmon.out will not be generated\n");
+        return;
+    }
+}
+
+void _monstartup(uintptr_t lowpc, uintptr_t highpc) {
```

Comment:
It looks like on POSIX at least, this is a *public function* which you can call to only profile a certain range of functions? https://man.freebsd.org/cgi/man.cgi?apropos=0&sektion=3&query=monstartup&manpath=FreeBSD+7.0-current&format=html

Is it as simple as just declaring this as public within the header to support this use-case? If so, I'd do it. If it's way more work than that, whatever, outside the scope of this PR, can come later. :)

Reply:
Didnt realize that monstartup should be public. Honestly, I dont see many people doing it that way since you still need portions of your code compiled with -pg to even produce a call graph.  I'll think about how I would support this use case.